src/realm/object-store/sync/impl/emscripten/socket_provider.cpp

```diff
@@ -0,0 +1,219 @@
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "socket_provider.hpp"
+#include <emscripten/websocket.h>
+
+using namespace realm;
+using namespace realm::_impl;
+using namespace realm::sync;
+
+namespace {
+struct EmscriptenTimer final : SyncSocketProvider::Timer {
+    using doubleMiliseconds = std::chrono::duration<double, std::chrono::milliseconds::period>;
+
+public:
+    EmscriptenTimer(std::chrono::milliseconds delay, SyncSocketProvider::FunctionHandler&& handler,
+                    util::EmscriptenScheduler& scheduler)
+        : m_handler(std::move(handler))
+        , m_timeout(emscripten_set_timeout(timeout_callback, doubleMiliseconds(delay).count(), this))
+        , m_scheduler(scheduler)
+    {
+    }
+    virtual ~EmscriptenTimer() final
+    {
+        cancel();
+    }
+    virtual void cancel() final
+    {
+        if (m_timeout && m_handler) {
+            emscripten_clear_timeout(*m_timeout);
+            m_scheduler.invoke([handler = std::move(m_handler)]() {
+                handler(Status(ErrorCodes::OperationAborted, "Timer canceled"));
+            });
+        }
+    }
+
+private:
+    SyncSocketProvider::FunctionHandler m_handler;
+    std::optional<int> m_timeout;
+    util::EmscriptenScheduler& m_scheduler;
+
+    static void timeout_callback(void* user_data)
+    {
+        auto timer = reinterpret_cast<EmscriptenTimer*>(user_data);
+        timer->m_timeout = std::nullopt;
+        timer->m_handler(Status::OK());
+    }
+};
+
+struct EmscriptenWebSocket final : public WebSocketInterface {
+public:
+    EmscriptenWebSocket(EMSCRIPTEN_WEBSOCKET_T socket, std::unique_ptr<WebSocketObserver> observer)
+        : m_socket(socket)
+        , m_sentinel(std::make_shared<LivenessSentinel>())
+        , m_observer(std::move(observer))
+    {
+        emscripten_websocket_set_onopen_callback(m_socket, m_observer.get(), open_callback);
+        emscripten_websocket_set_onmessage_callback(m_socket, m_observer.get(), message_callback);
+        emscripten_websocket_set_onerror_callback(m_socket, m_observer.get(), error_callback);
+        emscripten_websocket_set_onclose_callback(m_socket, m_observer.get(), close_callback);
+    }
+
+    virtual ~EmscriptenWebSocket() final
+    {
+        emscripten_websocket_close(m_socket, 0, nullptr);
+        emscripten_websocket_delete(m_socket);
+    }
```

Comment:
My assumption is that if the `EmscriptenWebSocket` is destroyed it's because the Sync Client released it and no longer cares about the callback. But resetting the sentinel makes sense in any case.